LX-1539 Add version.info to Linux upgrade image

dtzeng opened the pull request:
Added a version.info file in the upgrade image.
`DLPX_MIN_VERSION` (same as on Illumos) is for limiting how far a customer can jump with an upgrade (right now on Illumos we have it as 2 major releases)
`DLPX_MIN_REBOOTED_VERSION` is a new field that we will use to disallow deferred upgrades for a version.  Open to any better names for this..

Testing:
1. Ran `git ab-pre-push -v internal-dev`
2. Uploaded upgrade image
3. Verified version.info contents were correct
```
delphix@dt-linux:~$ cat /var/dlpx-update/9999.0.0.0/version.info 
#
# This file is used by app-gate to identify an image's version metadata
# after it is uploaded.
#
# Note that any changes to this file's format is a flag-day and must be
# coupled with an app-gate change.
#

# The application (product) version.
VERSION=2018.10.23.00

# The minimum version an engine must be on, in order to upgrade to this image.
MINIMUM_VERSION=5.3.0.0

# The minimum version an engine did a reboot upgrade to, in order to upgrade to this image.
MINIMUM_REBOOT_REQUIRED_VERSION=5.3.0.0
```

prakashsurya commented on live-build/misc/live-build-hooks/82-upgrade-repository.binary:
```diff
@@ -149,6 +149,22 @@ aptly repo add upgrade-repository ./*.deb
+#     upgrade to, in order to upgrade to this image.
+#
+rm -rf version.info
+cat <<-EOF >version.info
```
What do you think about creating a `template.info` file in `live-build/misc/live-build-hooks`, similar to `template.ovf` and `template.ctl` to store this file? Then, we can copy it and use `sed` to replace the `DLPX_VERSION` value, similar to what we do for `template.ctl` earlier in this script.

I ask because I wonder if that'd make it more clear/easier to maintain the versions that are infrequently changing. Having a separate "template" file might make it easier to point folks to it, to show them the current values, as well as make it easier for somebody not familiar with this repository to change them (e.g. a release lead).

What do you think?

pandar00 commented on live-build/misc/live-build-hooks/82-upgrade-repository.binary:
```diff
@@ -149,6 +149,22 @@ aptly repo add upgrade-repository ./*.deb
+#   DLPX_MIN_REBOOTED_VERSION: The minimum version an engine did a reboot
+#     upgrade to, in order to upgrade to this image.
+#
+rm -rf version.info
```
Why do we delete this beforehand? The subsequent `cat` operation already uses `>` to replace `version.info` if it already exists.

pandar00 commented on live-build/misc/live-build-hooks/82-upgrade-repository.binary:
```diff
@@ -149,6 +149,22 @@ aptly repo add upgrade-repository ./*.deb
+#
+# We also want to include a file about the current image's version metadata:
+#   DLPX_VERSION: The application (product) version.
+#   DLPX_MIN_VERSION: The minimum version an engine must be on, in order to
```
> Do we need to keep the DLPX_ prefixes?

I agree, we are working with Delphix appliance upgrade image always so `DLPX_` prefix seem redundant.

> I'd prefer to replace MIN with MINIMUM

I agree as well. I generally dislike abbreviations but `MIN` is fairly a standard so I also don't feel strongly.

In addition, I noticed we have `DLPX_FMT_VERSION` attribute used in `version.info` file. I'm not sure if that field is used to denote version format of the `version.info` file or the specific version of the upgrade image but if it's used for the former, then we could possibly leverage it to denote Illumos vs Linux style versioning based on some version.

However, I'm also fine as it is.

dtzeng commented:
@prakashsurya @pandar00 
- created template file
- updated comments to explain dependency with app-gate
- dropped DLPX prefixes
- renamed to `MINIMUM_REBOOT_REQUIRED_VERSION`.  I think I still like this better than `DLPX_MIN_VDB_ONLINE_VERSION` because this is controlling what the code has to do.  VDBs going down are just a fallout from needing to reboot (which may be necessary for other reasons).  And who knows, maybe we figure out how keep VDBs up during a reboot in the future! :smile:

prakashsurya commented:
Thanks, I’ll give this another look in the morning.

dtzeng commented on live-build/misc/live-build-hooks/template.info:
```diff
@@ -0,0 +1,16 @@
+MINIMUM_VERSION=5.3.0.0
+
+# The minimum version an engine did a reboot upgrade to, in order to upgrade to this image.
+MINIMUM_REBOOT_REQUIRED_VERSION=5.3.0.0
```
Thanks for the suggestion, I think your description is clearer.  Will update diff

dtzeng commented:
bors r+
